### PR TITLE
Cache instruments so repeatedly creating identical instruments doesn't leak memory

### PR DESCRIPTION
**Pull request by dashpole**
Resolve https://github.com/open-telemetry/opentelemetry-go/issues/3527
Fixes https://github.com/open-telemetry/opentelemetry-go-contrib/issues/4226.

This is one potential solution to the problem of creating multiple instances of an instrumentation library leaking memory.  Alternatively, we could require users to cache the instrumentation library instance itself, as was done in https://github.com/googleapis/google-api-go-client/pull/2329.

This works well for synchronous instruments, as making observations on the instrument from multiple instances of a library correctly aggregates across instances.  But for async instruments, this is a bit more confusing.  e.g. what should happen if I instantiate two instances of the runtime metrics library?  Presumably the last observation should win?

This also only solves the problem of caching instruments themselves.  The pattern of making multiple instances of an instrumentation library would still leak memory if each instance created and registered a new Callback.  If RegisterCallback is used, the callback can be `Unregistered()`, but if the callback is passed as an option to the instrument, it will exist forever, which would also leak memory.

TODO:

* Benchmarks
* Cache callbacks as well

### CODE REVIEW

**Comment by dashpole** [comment]
Ah, cool.  I'd definitely like feedback from @MadVikingGod, then

File: sdk/metric/meter.go [hunk 1]
```diff
@@ -108,6 +122,9 @@ func (m *meter) Int64Histogram(name string, options ...metric.Int64HistogramOpti
 // It registers callbacks for each reader's pipeline.
 func (m *meter) int64ObservableInstrument(id Instrument, callbacks []metric.Int64Callback) (int64Observable, error) {
 	inst := newInt64Observable(m, id.Kind, id.Name, id.Description, id.Unit)
+	// If we are re-using the instrument, measures have already been appended
+	// and we don't need to append them again.
+	shouldAppendMeasures := len(inst.measures) == 0
```
Comment (MrAlias) [inline sdk/metric/meter.go:127]: Won't this also apply if a view is applied to a different instrument?

Comment (dashpole) [reply sdk/metric/meter.go:127]: Not sure I understand.  This means either:

* This is the first time an instrument with that name/description/unit is created, thus the measures haven't been initialized yet, OR
* The instrument was previously created, but with a drop aggregation, so there weren't any measures. We will still have a drop aggregation, so in that case `shouldAppendMeasures` won't ever be checked below.

**Comment by MadVikingGod** [comment]
> Ah, cool. I'd definitely like feedback from @MadVikingGod, then

That work was pulled out because of added complexity, trying to get something working, and concerns over memory freeing.

At first glance, this looks to add caching in a simple way.  I would suggest putting a benchmark to ensure that it performs better when reused and doesn't affect single-use too much.

**Comment by MrAlias** [comment]
> It isn't clear if the aggregators cache is still needed if we are caching instruments.

I think this is still going to be needed to ensure view modified instruments that result in the same aggregators get the same aggregators. But I haven't tested it.

**Comment by MrAlias** [comment]
> The pattern of making multiple instances of an instrumentation library would still leak memory if each instance created and registered a new Callback.

This is a bit worrisome. Solving the issue to return the same instrument instance may lead users into thinking they can just make the same calls repeatedly.

Should we document that only first call using `With*Callback` will be honored? And users should use `RegisterCallback` after that?

**Comment by dashpole** [comment]
Looks like python logs a warning and ignores callbacks after the first: https://github.com/open-telemetry/opentelemetry-python/blob/975733c71473cddddd0859c6fcbd2b02405f7e12/opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/__init__.py#L171

**Comment by MrAlias** [comment]
This should resolve https://github.com/open-telemetry/opentelemetry-go/issues/3527

**Comment by MrAlias** [comment]
> Looks like python logs a warning and ignores callbacks after the first: https://github.com/open-telemetry/opentelemetry-python/blob/975733c71473cddddd0859c6fcbd2b02405f7e12/opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/__init__.py#L171

+1 to this behavior being implemented here.

File: sdk/metric/cache.go [hunk 2]
```diff
@@ -52,3 +52,50 @@ func (c *cache[K, V]) Lookup(key K, f func() V) V {
...
+	})
+	return combined.val, combined.err
+}
+
+// HasKey returns true if Lookup has previously been called with that key
+//
+// HasKey is safe to call concurrently.
+func (c *cacheWithErr[K, V]) HasKey(key K) bool {
+	return c.cache.HasKey(key)
+}
```
Comment (MrAlias) [inline sdk/metric/cache.go:101]: This doesn't need to be defined given `cache` is embedded in `cacheWithErr`, right?

**Comment by dashpole** [comment]
This now ignores callbacks after the first.

For benchmarking, If I run the existing `BenchmarkInstrumentCreation`, it shows a substantial increase in performance because it is now returning the cached version instead of making a new one each time:

From main
```
$ go test -bench=BenchmarkInstrumentCreation
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/sdk/metric
cpu: Intel(R) Xeon(R) CPU @ 2.20GHz
BenchmarkInstrumentCreation-24    	  104380	     11401 ns/op	    2370 B/op	      70 allocs/op
PASS
ok  	go.opentelemetry.io/otel/sdk/metric	1.384s
```

This PR: 
```
$ go test -bench=BenchmarkInstrumentCreation
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/sdk/metric
cpu: Intel(R) Xeon(R) CPU @ 2.20GHz
BenchmarkInstrumentCreation-24    	  318759	      3187 ns/op	     480 B/op	       6 allocs/op
PASS
ok  	go.opentelemetry.io/otel/sdk/metric	1.144s
```

If I move NewMeterProvider and Meter() inside the loop (so that it isn't using the cached one), I get

From main
```
$ go test -bench=BenchmarkInstrumentCreation
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/sdk/metric
cpu: Intel(R) Xeon(R) CPU @ 2.20GHz
BenchmarkInstrumentCreation-24    	   41757	     26801 ns/op	   13242 B/op	     148 allocs/op
PASS
ok  	go.opentelemetry.io/otel/sdk/metric	1.488s
```

This PR: 
```
$ go test -bench=BenchmarkInstrumentCreation
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/sdk/metric
cpu: Intel(R) Xeon(R) CPU @ 2.20GHz
BenchmarkInstrumentCreation-24    	   27579	     44411 ns/op	   23228 B/op	     237 allocs/op
PASS
ok  	go.opentelemetry.io/otel/sdk/metric	1.745s
```

Which is definitely slower, but that may be because we are doing some extra instantiation of the caches in the MeterProvider.

File: sdk/metric/cache.go (hunk 2 above)
Comment (dashpole) [reply sdk/metric/cache.go]: fixed

**Review by MrAlias — APPROVED** [review]

**Review by dmathieu — APPROVED** [review]

**Pull request merged**